File: src/construction.f
      subroutine mfunpl1(par,siq,grad,m,lpar,ngrad,pen,z,w,erg)
C  
C   only call for m>1  otherwise we have a trivial solution  c1 = 0, c2 = -log(mean(siq))
C   corresponding to an isotropic tensor and weight 1
C
C   code is restricted to m<=7 (six nontrivial compartments)
C
      implicit logical (a-z)
      integer m,lpar,ngrad
      real*8 par(lpar),siq(ngrad),grad(3,ngrad),z(ngrad,m),erg,pen
      integer i,j,i3,mode,r
      real*8 c1,w(ngrad),sw,sth,z1,p0,p1,d1,d2,d3,emc1,work(1000),s(7)
      c1 = par(1)
      emc1 = exp(-c1)
      DO j=1,ngrad
         z(j,1) = emc1
      END DO
      DO i = 2,m
         i3=2*i
         p0=par(i3-2)
         p1=par(i3-1)
         sth = sin(p0)
         d1 = sth*cos(p1)
         d2 = sth*sin(p1)
         d3 = cos(p0)
         DO j = 1,ngrad
            z1 = d1*grad(1,j)+d2*grad(2,j)+d3*grad(3,j)
            z(j,i) = exp(-c1*z1*z1)
         END DO
      END DO
C  
C    need to copy siq to w since corresponding argument of dgels 
C    will be overwritten by the solution
C
      call dcopy(ngrad,siq,1,w,1)
      call dgelss(ngrad,m,1,z,ngrad,w,ngrad,s,-1.d0,r,work,1000,mode)
      IF(mode.ne.0) THEN
         call intpr("mode",4,mode,1)
         erg = 1d20
      ELSE 
         IF(work(1).gt.1000) THEN
            call intpr("ngrad",5,ngrad,1)
            call intpr("m",1,m,1)
            call dblepr("optimal LWORK",13,work,1)
         END IF
         sw=0.d0
C penalize for extreme c1 values
         if(c1.gt.5.d0) sw=sw+c1-5.d0
C penalize for small or negative c1 values
         if(c1.lt.1.d-2) sw=sw-1.d3*c1+1.d1
         z1=0.d0
         DO i=1,m
            if(w(i).gt.0.d0) z1=z1+w(i)
         END DO
         DO i=1,m
            if(w(i).lt.z1/pen) sw=sw-pen/z1*w(i)+1.d0/z1
         END DO
         DO i=m+1,ngrad
            sw=sw+w(i)**2
         END DO
         IF(r.lt.m) THEN
            call intpr("rank",4,r,1)
            call dblepr("par",3,par,lpar)
            call dblepr("siq",3,siq,ngrad)
            DO i=1,m
               call dblepr("z",1,z(1,i),ngrad)
            END DO
            call dblepr("w",1,w,m)
            call dblepr("rss",3,sw,1)
         END IF
         erg=sw
      END IF
      call rchkusr()
      RETURN
      END 
C
C __________________________________________________________________
C
      subroutine mfunpl1g(par,siq,grad,m,lpar,ngrad,z,w,erg,
     1                    ztz,fv,dfv,qv,dqv,dh,dz,dztz,zts,
     1                    dzts,dw)
C  
C   only call for m>1  otherwise we have a trivial solution  c1 = 0, c2 = -log(mean(siq))
C   corresponding to an isotropic tensor and weight 1
C
C
C   code is restricted to m<=6
C
      implicit logical (a-z)
      integer m,mm1,lpar,ngrad
      real*8 par(lpar),siq(ngrad),grad(3,ngrad),z(ngrad,m),fv(ngrad),
     1   erg,dfv(lpar,ngrad),qv(m,ngrad),dqv(2,m,ngrad),dh(lpar),
     1   ztz(m,m),dz(ngrad,m,lpar),dztz(m,m,lpar),zts(m),dzts(m),
     1   dw(m,lpar)
      integer i,j,k,i3,ind(6),mode
      real*8 c1,w(m),sw,sth,cth,spsi,cpsi,z1,p0,p1,d1,d2,d3,emc1,
     1       work(300)
      c1 = par(1)
      emc1 = exp(-c1)
      DO j=1,ngrad
         z(j,1) = emc1
         dz(j,1,1) = -emc1
         DO k=2,lpar
            dz(j,1,k) = 0.d0
         END DO
      END DO
      DO i = 2,m
         i3=2*i
         p0=par(i3-2)
         p1=par(i3-1)
         sth = sin(p0)
         cth = cos(p0)
         spsi = sin(p1)
         cpsi = cos(p1)
         d1 = sth*cpsi
         d2 = sth*spsi
         d3 = cth
         DO j = 1,ngrad
            z1 = d1*grad(1,j)+d2*grad(2,j)+d3*grad(3,j)
            z(j,i) = exp(-c1*z1*z1)
            qv(i,j) = z1*z1
            dqv(1,i,j) = 2.d0*(cth*(cpsi*grad(1,j)+spsi*grad(2,j))-
     1                         sth*grad(3,j))*qv(i,j)
            dqv(2,i,j) = 2.d0*sth*(cpsi*grad(2,j)-spsi*grad(1,j))*
     1                         qv(i,j)
         END DO
      END DO
C
C   compute dz/dpar in dz  for i>1
C
      DO i=2,m
         DO j=1,ngrad
            dz(j,i,1) = -qv(i,j)*z(j,i)
            DO k=2,m
              dz(j,i,2*k-2) = -c1*dqv(1,k,j)*z(j,i)
              dz(j,i,2*k-1) = -c1*dqv(2,k,j)*z(j,i)
            END DO
         END DO
      END DO
C
C   compute t(z)%*%z
C
      call DSYRK('U','T',m,ngrad,1.d0,z,ngrad,0.d0,ztz,m)
C
C   invert ztz in place
C
      call DGETRI(m,ztz,m,ind,work,30,mode)
      IF(mode.ne.0) THEN
         call intpr("mode1",4,mode,1)
      END IF
C
C   ztz now contains (t(z)%*%z)^(-1)
C
C
C    compute t(z)%*%s
C
      
      
C  
C    siq will be replaced, need to copy it if C-version of optim is used
C
      call dcopy(ngrad,siq,1,w,1)
      call dgels('N',ngrad,m,1,z,ngrad,w,ngrad,work,30,mode)
      IF(mode.ne.0) THEN
         call intpr("mode2",4,mode,1)
      END IF
C
C   now compute what's needed for the gradient
C
      DO j = 1,ngrad
C        call intpr("j1",2,j,1)
         fv(j) = 0.d0 
         fv(j) = w(1)*emc1
         dfv(1,j) = -w(i)*emc1
         DO i = 2,m
            z1 = exp(-c1*qv(i,j))
            fv(j) = fv(j)+w(i)*z1
            dfv(1,j) = dfv(1,j)-w(i)*qv(i,j)*z1
            dfv(2*i-2,j) = -w(i)*c1*dqv(1,i,j)*z1
            dfv(2*i-1,j) = -w(i)*c1*dqv(2,i,j)*z1
         END DO
      END DO
C
C   now compute the gradient in dh
C
      DO i=1,lpar
         dh(i) = 0.d0
         DO j = 1,ngrad
           dh(i) = dh(i)-dfv(i,j)*(siq(j)-fv(j))
         END DO
         dh(i) = 2.d0*dh(i)
      END DO
      RETURN
      END 
C
C __________________________________________________________________
C
      subroutine mfunpl0g(par,siq,grad,m,lpar,ngrad,z,w,work1,erg,
     1                    fv,dfv,qv,dqv,dh)
      implicit logical (a-z)
      integer m,lpar,ngrad
      real*8 par(lpar),siq(ngrad),grad(3,ngrad),z(ngrad,m),erg,
     1   fv(ngrad),dfv(lpar,ngrad),qv(m,ngrad),dqv(2,m,ngrad),dh(lpar)
      integer i,j,i3,ind(10),mode
      real*8 c1,w(m),sw,sth,cth,spsi,cpsi,z1,p0,p1,d1,d2,d3,
     1       work1(ngrad),work2(10)
      c1 = par(1)
      sw = 0
      DO i = 1,m
         i3=2*i
         p0=par(i3)
         p1=par(i3+1)
         sth = sin(p0)
         cth = cos(p0)
         spsi = sin(p1)
         cpsi = cos(p1)
         d1 = sth*cpsi
         d2 = sth*spsi
         d3 = cth
         DO j = 1,ngrad
            z1 = d1*grad(1,j)+d2*grad(2,j)+d3*grad(3,j)
            z(j,i) = exp(-c1*z1*z1)
            qv(i,j) = z1*z1
            dqv(1,i,j) = 2.d0*(cth*(cpsi*grad(1,j)+spsi*grad(2,j))-
     1                         sth*grad(3,j))*qv(i,j)
            dqv(2,i,j) = 2.d0*sth*(cpsi*grad(2,j)-spsi*grad(1,j))*
     1                         qv(i,j)
         END DO
      END DO
C  
C    siq will be replaced, need to copy it if C-version of optim is used
C
      call dcopy(ngrad,siq,1,fv,1)
      call nnls(z,ngrad,ngrad,m,fv,w,erg,work2,work1,ind,mode)
      IF(mode.gt.1) THEN
         call intpr("mode",4,mode,1)
      END IF
C
C   now compute what's needed for the gradient
C
      DO j = 1,ngrad
C        call intpr("j1",2,j,1)
         fv(j) = 0.d0
         dfv(1,j) = 0.d0 
         DO i = 1,m
            z1 = exp(-c1*qv(i,j))
            fv(j) = fv(j)+w(i)*z1
            dfv(1,j) = dfv(1,j)-w(i)*qv(i,j)*z1
            dfv(2*i,j) = -w(i)*c1*dqv(1,i,j)*z1
            dfv(2*i+1,j) = -w(i)*c1*dqv(2,i,j)*z1
         END DO
      END DO
C
C   now compute the gradient in dh
C
      DO i=1,lpar
         dh(i) = 0.d0
         DO j = 1,ngrad
           dh(i) = dh(i)-dfv(i,j)*(siq(j)-fv(j))
         END DO
         dh(i) = 2.d0*dh(i)
      END DO
      RETURN
      END 
C
C __________________________________________________________________
C
      subroutine mfunpl(par,siq,grad,m,lpar,ngrad,z,w,work1,erg)
      implicit logical (a-z)
      integer m,lpar,ngrad
      real*8 par(lpar),siq(ngrad),grad(3,ngrad),z(ngrad,m),erg
      integer i,j,i3,ind(10),mode
      real*8 c1,w(m),sw,sth,z1,p0,p1,d1,d2,d3,
     1       work1(ngrad),work2(10)
C      c1 = exp(par(1))
      c1 = par(1)
      sw = 0
      DO i = 1,m
         i3=2*i
         p0=par(i3)
         p1=par(i3+1)
         sth = sin(p0)
         d1 = sth*cos(p1)
         d2 = sth*sin(p1)
         d3 = cos(p0)
         DO j = 1,ngrad
            z1 = d1*grad(1,j)+d2*grad(2,j)+d3*grad(3,j)
            z(j,i) = exp(-c1*z1*z1)
         END DO
      END DO
C  
C    siq will be replaced, need to copy it if C-version of optim is used
C
      call nnls(z,ngrad,ngrad,m,siq,w,erg,work2,work1,ind,mode)
      IF(mode.gt.1) THEN
         call intpr("mode",4,mode,1)
      END IF
      RETURN
      END 
C
C __________________________________________________________________
C
      subroutine mfunplgr(par,siq,grad,m,lpar,ngrad,z,w,work1,erg,
     1                    fv,dfv,qv,dqv,dh)
      implicit logical (a-z)
      integer m,lpar,ngrad
      real*8 par(lpar),siq(ngrad),grad(3,ngrad),z(ngrad,m),erg,
     1   fv(ngrad),dfv(lpar,ngrad),qv(m,ngrad),dqv(2,m,ngrad),dh(lpar)
      integer i,j,i3,ind(10),mode
      real*8 c1,w(m),sw,sth,cth,spsi,cpsi,z1,p0,p1,d1,d2,d3,
     1       work1(ngrad),work2(10)
      c1 = par(1)
      sw = 0
      DO i = 1,m
         i3=2*i
         p0=par(i3)
         p1=par(i3+1)
         sth = sin(p0)
         cth = cos(p0)
         spsi = sin(p1)
         cpsi = cos(p1)
         d1 = sth*cpsi
         d2 = sth*spsi
         d3 = cth
         DO j = 1,ngrad
            z1 = d1*grad(1,j)+d2*grad(2,j)+d3*grad(3,j)
            z(j,i) = exp(-c1*z1*z1)
            qv(i,j) = z1*z1
            dqv(1,i,j) = 2.d0*(cth*(cpsi*grad(1,j)+spsi*grad(2,j))-
     1                         sth*grad(3,j))*qv(i,j)
            dqv(2,i,j) = 2.d0*sth*(cpsi*grad(2,j)-spsi*grad(1,j))*
     1                         qv(i,j)
         END DO
      END DO
C  
C    siq will be replaced, need to copy it if C-version of optim is used
C
      call dcopy(ngrad,siq,1,fv,1)
      call nnls(z,ngrad,ngrad,m,fv,w,erg,work2,work1,ind,mode)
      IF(mode.gt.1) THEN
         call intpr("mode",4,mode,1)
      END IF
C
C   now compute what's needed for the gradient
C
      DO j = 1,ngrad
C        call intpr("j1",2,j,1)
         fv(j) = 0.d0
         dfv(1,j) = 0.d0 
         DO i = 1,m
            z1 = exp(-c1*qv(i,j))
            fv(j) = fv(j)+w(i)*z1
            dfv(1,j) = dfv(1,j)-w(i)*qv(i,j)*z1
            dfv(2*i,j) = -w(i)*c1*dqv(1,i,j)*z1
            dfv(2*i+1,j) = -w(i)*c1*dqv(2,i,j)*z1
         END DO
      END DO
C
C   now compute the gradient in dh
C
      DO i=1,lpar
         dh(i) = 0.d0
         DO j = 1,ngrad
           dh(i) = dh(i)-dfv(i,j)*(siq(j)-fv(j))
         END DO
         dh(i) = 2.d0*dh(i)
      END DO
      RETURN
      END 
C
C __________________________________________________________________
C
      subroutine mfunpl2(par,siq,grad,m,p,lpar,ngrad,z,w,work1,erg)
      implicit logical (a-z)
      integer m,lpar,ngrad
      real*8 par(lpar),siq(ngrad),grad(3,ngrad),z(ngrad,m),erg,p
      integer i,j,i3,ind(10),mode
      real*8 c1,c2,w(m),sw,sth,z1,p0,p1,d1,d2,d3,
     1       work1(ngrad),work2(10)

      c1 = exp(par(1))
      c2 = exp(par(2))
      DO i = 1,m
         i3=2*i+1
         p0=par(i3)
         p1=par(i3+1)
         sth = sin(p0)
         d1 = sth*cos(p1)
         d2 = sth*sin(p1)
         d3 = cos(p0)
         DO j = 1,ngrad
            z1 = d1*grad(1,j)+d2*grad(2,j)+d3*grad(3,j)
            z(j,i) = exp(-p*log(1.d0+(c2+c1*z1*z1)/p))
         END DO
      END DO
C  
C    siq will be replaced, need to copy it if C-version of optim is used
C
      call nnls(z,ngrad,ngrad,m,siq,w,erg,work2,work1,ind,mode)
      IF(mode.gt.1) THEN
         call intpr("mode",4,mode,1)
      END IF
      sw = -1.d0
      DO i=1,m
         sw=sw+w(i)
      END DO
C add penalty for sum of weight .neq. 1
      erg=erg+1.d2*sw*sw
      RETURN
      END 

File: R/construction.r
mfunpl1 <- function(par,siq,grad,pen=1e2){
#
#   evaluate rss for Mixtensor-model (with isotropic component)
#
lpar <- length(par)
m <- (lpar+1)/2
ngrad <- dim(grad)[1]
.Fortran("mfunpl1",as.double(par),#par(lpar)
                as.double(siq),#siq(ngrad)
                as.double(t(grad)),#grad(3,ngrad)
                as.integer(m),#number of components
                as.integer(lpar),#number of parameters
                as.integer(ngrad),#number of gradients
                as.double(pen),#penalty for negative weights
                double(ngrad*m),#z(ngrad,m) working array
                double(ngrad),#w(ngrad) working array
                erg = double(1),#residual sum of squares
                PACKAGE="dti")$erg
}
gmfunpl1 <- function(par,siq,grad){
#
#   evaluate rss for Mixtensor-model
#
lpar <- length(par)
m <- (lpar+1)/2
ngrad <- dim(grad)[1]
.Fortran("mfunpl1g",as.double(par),#par
                as.double(siq),#siq
                as.double(t(grad)),#grad
                as.integer(m),#m
                as.integer(lpar),#lpar
                as.integer(ngrad),#ngrad
                double(ngrad*m),#z
                double(m),#w
                double(ngrad),#work
                double(1),#erg
                double(ngrad),#fv
                double(lpar*ngrad),#dfv
                double(m*ngrad),#qv
                double(2*m*ngrad),#dqv
                gradient=double(lpar),#dh
                PACKAGE="dti")$gradient
}
mfunplwghts1 <- function(par,siq,grad,pen=1e2){
#
#   get weights for Mixtensor-model (with isotropic component) and extract parameters
#
lpar <- length(par)
m <- (lpar+1)/2
ngrad <- dim(grad)[1]
w<-.Fortran("mfunpl1",as.double(par),#par(lpar)
                as.double(siq),#siq(ngrad)
                as.double(t(grad)),#grad(3,ngrad)
                as.integer(m),#number of components
                as.integer(lpar),#number of parameters
                as.integer(ngrad),#number of gradients
                as.double(pen),#penalty for negative weights
                double(ngrad*m),#z(ngrad,m) working array
                w=double(ngrad),#w(ngrad) working array
                double(1),#residual sum of squares
                PACKAGE="dti")$w[1:m]
           wm1 <- w[-1] # first comonent corresponds to isotropic part
           o <- order(wm1,decreasing=TRUE)
           ord <- sum(wm1>0)
           if(ord<m-1){
              o <- o[1:ord]
           }
           sw <- sum(w[w>0])
           lev <- c(par[1],-log(sw))
           if(ord>0){
           mix <- wm1[o]/sw
           } else {
           mix <- NULL
           } 
           or <- matrix(par[2:lpar],2,m-1)[,o,drop=FALSE]
           or[1,or[1,]<0] <- or[1,or[1,]<0]+pi
           or[1,or[1,]>pi] <- or[1,or[1,]>pi]-pi
           or[2,or[2,]<0] <- or[2,or[2,]<0]+2*pi
           or[2,or[2,]>2*pi] <- or[2,or[2,]>2*pi]-2*pi
           par <- c(par[1],or[,1:ord])
list(ord=ord,lev=lev,mix=mix,orient=or,par=par)
}
mfunpl <- function(par,siq,grad){
#
#   evaluate rss for Mixtensor-model
#
lpar <- length(par)
m <- (lpar-1)/2
ngrad <- dim(grad)[1]
.Fortran("mfunpl",as.double(par),
                as.double(siq),
                as.double(t(grad)),
                as.integer(m),
                as.integer(lpar),
                as.integer(ngrad),
                double(ngrad*m),
                double(m),
                double(ngrad),
                erg = double(1),
                PACKAGE="dti")$erg
}
gmfunpl <- function(par,siq,grad){
#
#   evaluate rss for Mixtensor-model
#
lpar <- length(par)
m <- (lpar-1)/2
ngrad <- dim(grad)[1]
.Fortran("mfunplgr",as.double(par),#par
                as.double(siq),#siq
                as.double(t(grad)),#grad
                as.integer(m),#m
                as.integer(lpar),#lpar
                as.integer(ngrad),#ngrad
                double(ngrad*m),#z
                double(m),#w
                double(ngrad),#work
                double(1),#erg
                double(ngrad),#fv
                double(lpar*ngrad),#dfv
                double(m*ngrad),#qv
                double(2*m*ngrad),#dqv
                gradient=double(lpar),#dh
                PACKAGE="dti")$gradient
}
mfunplwghts <- function(par,siq,grad){
#
#   get weights for Mixtensor-model and extract parameters
#
lpar <- length(par)
m <- (lpar-1)/2
ngrad <- dim(grad)[1]
w<-.Fortran("mfunpl",as.double(par),
                as.double(siq),
                as.double(t(grad)),
                as.integer(m),
                as.integer(lpar),
                as.integer(ngrad),
                double(ngrad*m),
                w=double(m),
                double(ngrad),
                erg = double(1),
                PACKAGE="dti")$w
           o <- order(w,decreasing=TRUE)
           ord <- sum(w>0)
           if(ord<m){
              o <- o[1:ord]
           }
           sw <- sum(w)
           lev <- c(par[1],-log(sw))
           mix <- (w[o]/sw)
           or <- matrix(par[2:lpar],2,m)[,o,drop=FALSE]
           or[1,or[1,]<0] <- or[1,or[1,]<0]+pi
           or[1,or[1,]>pi] <- or[1,or[1,]>pi]-pi
           or[2,or[2,]<0] <- or[2,or[2,]<0]+2*pi
           or[2,or[2,]>2*pi] <- or[2,or[2,]>2*pi]-2*pi
           par <- c(par[1],or[,1:ord])
list(ord=ord,lev=lev,mix=mix,orient=or,par=par)
}
mfunpl2 <- function(par,siq,grad,pex){
#
#   evaluate rss for Jian-model
#
lpar <- length(par)
m <- (lpar-2)/2
ngrad <- dim(grad)[1]
.Fortran("mfunpl2",as.double(par),
                as.double(siq),
                as.double(t(grad)),
                as.integer(m),
                as.double(pex),
                as.integer(lpar),
                as.integer(ngrad),
                double(ngrad*m),
                double(m),
                double(ngrad),
                erg = double(1),
                PACKAGE="dti")$erg
}
mfunpl2wghts <- function(par,siq,grad,pex){
#
#   get weights for Jian-model and extract parameters
#
lpar <- length(par)
m <- (lpar-2)/2
ngrad <- dim(grad)[1]
w<-.Fortran("mfunpl2",as.double(par),
                as.double(siq),
                as.double(t(grad)),
                as.integer(m),
                as.double(pex),
                as.integer(lpar),
                as.integer(ngrad),
                double(ngrad*m),
                w=double(m),
                double(ngrad),
                erg = double(1),
                PACKAGE="dti")$w
           o <- order(w,decreasing=TRUE)
           ord <- sum(w>0)
           if(ord<m){
              o <- o[1:ord]
           }
           lev <- exp(c(par[1],par[2]))
           mix <- if(ord==0) NULL else w[o]
           or <- matrix(par[3:lpar],2,m)[,o,drop=FALSE]
           or[1,or[1,]<0] <- or[1,or[1,]<0]+pi
           or[1,or[1,]>pi] <- or[1,or[1,]>pi]-pi
           or[2,or[2,]<0] <- or[2,or[2,]<0]+2*pi
           or[2,or[2,]>2*pi] <- or[2,or[2,]>2*pi]-2*pi
           par <- c(par[1],par[2],or[,1:ord])
list(ord=ord,lev=lev,mix=mix,orient=or,par=par)
}


dwiMixtensor.new <- function(object, ...) cat("No dwiMixtensor.new calculation defined for this class:",class(object),"\n")

setGeneric("dwiMixtensor.new", function(object,  ...) standardGeneric("dwiMixtensor.new"))

setMethod("dwiMixtensor.new",
          "dtiData",
          function(object,
                   maxcomp = 3,
                   maxneighb = 7,
                   reltol = 1e-8,
                   maxit = 5000,
                   optmethod = "Nelder-Mead")
           {
             cat("entering method dwiMixtensor\n")

             if (maxcomp > 4) stop("maximum number of tensor components is 4\n") # otherwise change number of parameters in C code!

             args <- sys.call(-1)
             args <- c(object@call, args)
             ngrad <- object@ngrad
             ddim <- object@ddim
             s0ind <- object@s0ind
             ns0 <- length(s0ind)

             cat("determine outliers ... ")
             z <- .Fortran("outlier",
                           as.double(object@si),
                           as.integer(prod(ddim)),
                           as.integer(ngrad),
                           as.logical((1:ngrad)%in%s0ind),
                           as.integer(ns0),
                           si      = integer(prod(ddim)*ngrad),
                           index   = integer(prod(ddim)),
                           lindex  = integer(1),
                           DUP     = FALSE,
                           PACKAGE = "dti")[c("si","index","lindex")]
             si <- array(z$si, c(ddim, ngrad))
             index <- if (z$lindex>0) z$index[1:z$lindex] else numeric(0)
             rm(z)
             cat("done\n")

             cat("prepare data and initial estimates ... ")
             # prepare data for optim
             s0 <- si[,,,s0ind,drop=FALSE]
             if (length(s0ind)>1) s0 <- apply(s0, 1:3, mean) else dim(s0) <- dim(s0)[1:3]
             # normalized DW data
             mask <- s0 > object@level
             siq <- si[,,,-s0ind,drop=FALSE]
             dim(siq) <- c(prod(ddim),ngrad-ns0)
             siq[mask,] <- sweep(siq[mask,],1,s0[mask],"/")
             dim(siq) <- c(ddim,ngrad-ns0)
             # heuristics to avoid DWI that are larger than s0.
             siqmed <- apply(siq, 1:3, median)
             siqmed[siqmed < .9] <- .9
             siqmed[siqmed > .99] <- .99
             siq <- sweep(siq,1:3,siqmed,pmin)
             # mask for calculation
             mask <- s0 > object@level
             grad <- t(object@gradient[,-s0ind])
             siind <- getsiind(siq, mask, grad, maxcomp, maxc=.866)
             # siind[1,,,] contains number of potential directions 
             # siind[-1,,,] contains indices of grad corresponding to these directions
             cat("done\n")

             cat("optimizing ... ")
             mm <- switch(optmethod, "Nelder-Mead" = 1,
                                     "BFGS"        = 3,
                                     1)
             pl <- 1

             # perform voxelwise optimization and order selection of tensor mixture model
             a <- .C("mixturepl",
                     as.integer(mm),                         # select mixture method
                     as.integer(prod(ddim)),                 # number of voxels
                     as.integer(mask),                       # calculation mask
                     as.double(siq),                         # DWI without s0
                     as.integer(siind),                      # DWI indices of local minima
                     as.integer(ngrad - length(s0ind)),      # number of DWI
                     as.double(grad),                        # gradient directions
                     as.integer(maxcomp),                    # max number of gradient neighbors
                     as.integer(pl),                         # exp for Jian model
                     as.integer(maxit),                      # max number of iterations for optim
                     as.double(reltol),                      # reltol crit for optim
                     order   = integer(prod(ddim)),           # selected order of mixture
                     lev     = double(2*prod(ddim)),         # logarithmic eigenvalues
                     mix     = double(maxcomp*prod(ddim)),   # mixture weights
                     orient  = double(2*maxcomp*prod(ddim)), # phi/theta for all mixture tensors
                     sigma2  = double(prod(ddim)),           # parameter variance ???
                     DUP     = FALSE,
                     PACKAGE = "dti")[c("order", "lev", "mix", "orient", "sigma2")]

             # set dimension attr
             dim(a$order) <- ddim;
             dim(a$lev) <- c(2, ddim);
             dim(a$mix) <- c(maxcomp, ddim);
             dim(a$orient) <- c(2, maxcomp, ddim);
             dim(a$sigma2) <- ddim;
             cat("done\n")

             # create and return new object
             invisible(new("dwiMixtensor",
                           call        = args,
                           ev          = a$lev,
                           mix         = a$mix,
                           orient      = a$orient,
                           order       = a$order,
                           p           = pl,
                           th0         = s0,
                           sigma       = a$sigma2,
                           scorr       = array(1, c(1,1,1)), # ???
                           bw          = c(0,0,0),           # ???
                           mask        = mask,
                           hmax        = 1,                  # ???
                           gradient    = object@gradient,
                           btb         = object@btb,
                           ngrad       = object@ngrad,
                           s0ind       = object@s0ind,
                           replind     = object@replind,
                           ddim        = object@ddim,
                           ddim0       = object@ddim0,
                           xind        = object@xind,
                           yind        = object@yind,
                           zind        = object@zind,
                           voxelext    = object@voxelext,
                           level       = object@level,
                           orientation = object@orientation,
                           source      = object@source,
                           outlier     = index,
                           scale       = 1,
                           method      = "mixtensor")
                       )
           })
